File: apps/hyperliquid-recorder/tests/test_health.py
from __future__ import annotations

import time

from hyperliquid_recorder.health import HealthState


def test_health_requires_clickhouse_and_market_freshness() -> None:
    state = HealthState(expected_coins=("BTC", "ETH"), stale_seconds=30)
    assert not state.is_ready()

    state.set_clickhouse_ok(True)
    assert not state.is_ready()

    state.set_market_seen("BTC")
    state.set_market_seen("ETH")
    assert state.is_ready()


def test_health_becomes_unready_when_market_stale() -> None:
    state = HealthState(expected_coins=("BTC",), stale_seconds=0)
    state.set_clickhouse_ok(True)
    state.set_market_seen("BTC")
    time.sleep(0.01)
    assert not state.is_ready()

File: apps/ondo-recorder/tests/test_health.rs
use std::time::Duration;

use ondo_recorder::health::{HealthState, Market};

#[test]
fn test_health_requires_clickhouse_and_market_freshness() {
    let state = HealthState::new(
        vec![
            Market::new("AAPLon", "ethereum-1"),
            Market::new("NVDAon", "ethereum-1"),
        ],
        30,
    );
    assert!(!state.is_ready());

    state.set_clickhouse_ok(true);
    assert!(!state.is_ready());

    state.set_market_seen("AAPLon", "ethereum-1");
    state.set_market_seen("NVDAon", "ethereum-1");
    assert!(state.is_ready());
}

#[test]
fn test_health_tracks_same_symbol_across_chains_independently() {
    let state = HealthState::new(
        vec![
            Market::new("AAPLon", "ethereum-1"),
            Market::new("AAPLon", "bsc-56"),
        ],
        30,
    );
    state.set_clickhouse_ok(true);
    state.set_market_seen("AAPLon", "ethereum-1");
    assert!(
        !state.is_ready(),
        "one chain still missing a poll should keep state unready"
    );
    state.set_market_seen("AAPLon", "bsc-56");
    assert!(state.is_ready());
}

#[test]
fn test_health_becomes_unready_when_market_stale() {
    let state = HealthState::new(vec![Market::new("AAPLon", "ethereum-1")], 0);
    state.set_clickhouse_ok(true);
    state.set_market_seen("AAPLon", "ethereum-1");
    std::thread::sleep(Duration::from_millis(10));
    assert!(!state.is_ready());
}
